packages/blade/src/components/Dropdown/dropdownUtils.ts

```diff
@@ -0,0 +1,295 @@
+
+  // first check if there is an exact match for the typed string
+  if (firstMatch) {
+    return options.indexOf(firstMatch);
+  }
+
+  // if the same letter is being repeated, cycle through first-letter matches
+  else if (allSameLetter(filter.split(''))) {
+    const matches = filterOptions(orderedOptions, filter[0]);
+    return options.indexOf(matches[0]);
+  }
+
+  // if no matches, return -1
+  else {
+    return -1;
+  }
+}
+
+/**
+ * This functions makes sure the optionsIndex is not going out of possible options
+ */
+export function getUpdatedIndex(
+  currentIndex: number,
+  maxIndex: number,
+  action: SelectActionsType,
+): number {
+  // On PageUP or PageDown, we jump focus by 10 items or to the first or last element
+  // Details: https://www.w3.org/WAI/ARIA/apg/example-index/combobox/combobox-select-only.html#:~:text=PageUp,to%20last%20option).
+  const pageSize = 10;
+
+  switch (action) {
+    case SelectActions.First:
+      return 0;
+    case SelectActions.Last:
+      return maxIndex;
+    case SelectActions.Previous:
+      return Math.max(0, currentIndex - 1);
+    case SelectActions.Next:
+      return Math.min(maxIndex, currentIndex + 1);
+    case SelectActions.PageUp:
+      return Math.max(0, currentIndex - pageSize);
+    case SelectActions.PageDown:
+      return Math.min(maxIndex, currentIndex + pageSize);
+    default:
+      return currentIndex;
+  }
+}
+
+/**
+ * Checks if the given HTML element is visible on screen
+ */
+export function isElementVisibleOnScreen(element: HTMLElement): boolean {
+  const bounding = element.getBoundingClientRect();
+
+  return (
+    bounding.top >= 0 &&
+    bounding.left >= 0 &&
+    bounding.bottom <= (window.innerHeight || document.documentElement.clientHeight) &&
+    bounding.right <= (window.innerWidth || document.documentElement.clientWidth)
+  );
+}
+
+/**
+ * Checks if the dropdown is scrollable
+ */
+export function isScrollable(element: HTMLElement): boolean {
+  return element && element.clientHeight < element.scrollHeight;
+}
+
+type ActionsType = {
+  setIsOpen: DropdownContextType['setIsOpen'];
+  selectCurrentOption: () => void;
+  onOptionChange: (action: SelectActionsType) => void;
+  onComboType: (letter: string, action: SelectActionsType) => void;
+};
+/**
+ * Performs the action when actionType is passed
+ *
+ * This function handles all the keydown actions.
+ */
+export const performAction = (
+  action: SelectActionsType,
+  e: FormInputOnKeyDownEvent,
+  actions: ActionsType,
+): boolean => {
+  const { event } = e;
+
+  switch (action) {
+    case SelectActions.Last:
+    // @ts-expect-error: intentional fallthrough, ignoring the warning
+    case SelectActions.First:
+      actions.setIsOpen(true);
+    // intentional fallthrough
+    case SelectActions.Next:
+    case SelectActions.Previous:
+    case SelectActions.PageUp:
+    case SelectActions.PageDown:
+      event.preventDefault();
+      actions.onOptionChange(action);
```

Comment:
what does this do? you call the option is changed with action arg?

Reply:
Inside onOptionChange event, we want to know which action triggered that (whether it was because of keyboard event, or mouse click) so just passing down that action here.